File: ios/Classes/TiAnimationAnimationView.m
/**
 * Ti.Lottie
 * Copyright (c) 2017-present by Hans Knöchel. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#import "TiAnimationAnimationView.h"
#import "TiAnimationAnimationViewProxy.h"

#import <Lottie/Lottie.h>

@implementation TiAnimationAnimationView

#pragma mark Internals

- (LOTAnimationView *)animationView {
  if (_animationView == nil) {
    id file = [[self proxy] valueForKey:@"file"];
    id autoStart = [[self proxy] valueForKey:@"autoStart"];
    id contentMode = [[self proxy] valueForKey:@"contentMode"];

    ENSURE_TYPE(file, NSString);
    ENSURE_TYPE_OR_NIL(autoStart, NSNumber);
    ENSURE_TYPE_OR_NIL(contentMode, NSNumber);

    // Handle both "file.json" and "file"
    if ([file hasSuffix:@".json"]) {
      file = [file stringByDeletingPathExtension];
    }

    _animationView = [LOTAnimationView animationFromJSON:[self loadAnimationFromJSON:file]];

    // Enable click-events
    UITapGestureRecognizer *tapGestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(didClickView:)];
    [_animationView addGestureRecognizer:tapGestureRecognizer];

    // Handle content mode
    NSArray<NSNumber *> *validContentModes = @[ @(UIViewContentModeScaleAspectFit), @(UIViewContentModeScaleAspectFill), @(UIViewContentModeScaleToFill) ];

    if (contentMode && [validContentModes containsObject:contentMode]) {
      [_animationView setContentMode:[TiUtils intValue:contentMode]];
    } else {
      [_animationView setContentMode:UIViewContentModeScaleAspectFit];

      if (contentMode) {
        NSLog(@"[ERROR] The \"contentMode\" property is not valid (CONTENT_MODE_ASPECT_FIT, CONTENT_MODE_ASPECT_FILL or CONTENT_MODE_SCALE_FILL), defaulting to CONTENT_MODE_ASPECT_FIT");
      }
    }

    [_animationView setFrame:[self bounds]];

    [self addSubview:_animationView];

    if ([TiUtils boolValue:autoStart def:NO]) {
      [[self animationView] play];
    }
  }

  return _animationView;
}

- (void)didClickView:(UIGestureRecognizer *)sender {
  if ([[self proxy] _hasListeners:@"click"]) {
    [[self proxy] fireEvent:@"click"];
  }
}

- (NSDictionary *)loadAnimationFromJSON:(NSString *)file {
  NSString *filePath = [[NSBundle mainBundle] pathForResource:[[self proxy] valueForKey:@"file"] ofType:nil inDirectory:nil];
  NSData *data = [NSData dataWithContentsOfFile:filePath];

  if (!data) {
    [self log:[NSString stringWithFormat:@"The specified file %@ could not be loaded.", file] forLevel:@"error"];
    return nil;
  }

  return [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:nil];
}

- (void)log:(NSString *)string forLevel:(NSString *)level {
  DebugLog(@"[%@] %@: %@", [level uppercaseString], NSStringFromClass([self class]), string);
}

#pragma mark Public APIs

- (void)playWithCompletionHandler:(KrollCallback *)callback {
  [[self animationView] playWithCompletion:^(BOOL animationFinished) {
    [self processCompleteEventWith:callback animationFinished:animationFinished];
  }];
}

- (void)playFromFrame:(NSNumber *)fromFrame toFrame:(NSNumber *)toFrame completion:(KrollCallback *)callback {
  [[self animationView] playFromFrame:fromFrame
                              toFrame:toFrame
                       withCompletion:^(BOOL animationFinished) {
                         [self processCompleteEventWith:callback animationFinished:animationFinished];
                       }];
}

// TODO: Expose to module in next major, because we need to break the 3 parameters into a dictionary to be more flexible
- (void)playFromProgress:(NSNumber *)fromProgress toProgress:(NSNumber *)toProgress completion:(KrollCallback *)callback {
  [[self animationView] playFromProgress:fromProgress.floatValue
                              toProgress:toProgress.floatValue
                          withCompletion:^(BOOL animationFinished) {
                            [self processCompleteEventWith:callback animationFinished:animationFinished];
                          }];
}

- (void)pause {
  [[self animationView] pause];
}

- (void)stop {
  [[self animationView] stop];
}

- (void)addView:(UIView *)view toKeypathLayer:(NSString *)layer {
  [[self animationView] addSubview:view toKeypathLayer:[LOTKeypath keypathWithString:layer]];
}

- (BOOL)isPlaying {
  return [[self animationView] isAnimationPlaying];
}

- (CGFloat)duration {
  return [[self animationView] animationDuration];
}

- (CGFloat)progress {
  return [[self animationView] animationProgress];
}

- (CGFloat)speed {
  return [[self animationView] animationSpeed];
}

- (BOOL)loop {
  return [[self animationView] loopAnimation];
}

- (void)setProgress:(CGFloat)progress {
  [[self animationView] setAnimationProgress:progress];
}

- (void)setSpeed:(CGFloat)speed {
  [[self animationView] setAnimationSpeed:speed];
}

- (void)setLoop:(BOOL)loop {
  [[self animationView] setLoopAnimation:loop];
}

- (void)setCacheEnabled:(BOOL)cacheEnabled {
  [[self animationView] setCacheEnable:cacheEnabled];
}

- (BOOL)cacheEnabled {
  return [[self animationView] cacheEnable];
}

- (void)processCompleteEventWith:(KrollCallback *)callback animationFinished:(BOOL)animationFinished {
  if ([[self proxy] _hasListeners:@"complete"]) {
    [[self proxy] fireEvent:@"complete" withObject:@{ @"animationFinished" : @(YES) }];
  }

  if (callback == nil) {
    return;
  }

  NSDictionary *propertiesDict = @{ @"finished" : @(animationFinished) };
  NSArray *invocationArray = [[NSArray alloc] initWithObjects:&propertiesDict count:1];

  [callback call:invocationArray thisObject:[self proxy]];
}

#pragma mark Layout utilities

#ifdef TI_USE_AUTOLAYOUT
- (void)initializeTiLayoutView {
  [super initializeTiLayoutView];
  [self setDefaultHeight:TiDimensionAutoFill];
  [self setDefaultWidth:TiDimensionAutoFill];
}
#endif

#pragma mark Layout helper

- (void)frameSizeChanged:(CGRect)frame bounds:(CGRect)bounds {
  for (UIView *child in [self subviews]) {
    [TiUtils setView:child positionRect:bounds];
  }

  [super frameSizeChanged:frame bounds:bounds];
}

@end

File: ios/Classes/TiAnimationAnimationView.swift
//
//  TiAnimationAnimationView.swift
//  ti.animation
//
//  Created by Hans Knöchel
//  Copyright (c) 2022 Hans Knöchel. All rights reserved.
//

import TitaniumKit
import Lottie

@objc(TiAnimationAnimationView)
public class TiAnimationAnimationView : TiUIView {

  var _animationView: AnimationView!
  
  private func animationView() -> AnimationView {
    if _animationView == nil {
      let file = TiUtils.stringValue(proxy.value(forKey: "file"))
      let jsonString = TiUtils.stringValue(proxy.value(forKey: "jsonString"))
      let autoStart = (proxy as! TiAnimationAnimationViewProxy).autoStart
      let speed = TiUtils.floatValueSwift((proxy as! TiAnimationAnimationViewProxy).speed)
      let loop = (proxy as! TiAnimationAnimationViewProxy).loop
      let contentMode: Int? = Int(TiUtils.intValue(proxy.value(forKey: "contentMode")))

      // Case A: Handle JSON file references (either from the packaged app or from the application cache
      if let file = file, file.hasSuffix(".json") {
        guard let dictionary = loadAnimationFrom(JSON: file) else {
          fatalError("Cannot find file!")
        }
        
        _animationView = AnimationView(animation: try! Animation(dictionary: dictionary))
      // Case B: Handle JSON strings
      } else if let jsonString = jsonString {
        let data = jsonString.data(using: .utf8)
        if let data = data, let jsonData = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
          _animationView = AnimationView(animation: try! Animation(dictionary: jsonData))
        }
      }
            
      // Enable click-events
      let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(didClickView))
      _animationView.addGestureRecognizer(tapGestureRecognizer)
      
      // Handle content mode
      if let contentMode = contentMode {
        let validContentModes: [UIView.ContentMode] = [.scaleAspectFit, .scaleAspectFill]
        if let nativeContentMode = UIView.ContentMode(rawValue: contentMode), validContentModes.contains(nativeContentMode) {
          _animationView.contentMode = nativeContentMode
        }
      } else {
        _animationView.contentMode = .scaleAspectFit
      }
      
      if let speed = speed {
        setSpeed(speed: speed)
      }
      
      _animationView.frame = bounds
      addSubview(_animationView)
      
      // Creation only: Loop mode
      if (loop) {
        _animationView.loopMode = .loop
      }
      
      // Creation only: Auto start
      if autoStart {
        play(with: nil)
      }
    }

    return _animationView
  }
 
  @objc
  private func didClickView(_ sender: UIGestureRecognizer) {
    guard let proxy = proxy else {
      return
    }

    if proxy._hasListeners("click") {
      proxy.fireEvent("click")
    }
  }
  
  private func processCompleteEvent(with callback: KrollCallback?, animationFinished: Bool) {
    guard let proxy = proxy else {
      return
    }

    if proxy._hasListeners("complete") {
      proxy.fireEvent("complete", with: ["animatonFinished": true])
    }
    
    guard let callback = callback else {
      return
    }
    
    let propertiesDict = ["finished": animationFinished]
    callback.call([propertiesDict], thisObject: proxy)
  }
  
  private func loadAnimationFrom(JSON file: String) -> [String: Any]? {
    var data: Data? = nil
    var fileString = file
    
    // Handle file document data directory
    if let documentsDirectoryURL = URL(string: file) {
      data = try? Data(contentsOf: documentsDirectoryURL, options: .mappedIfSafe)
    }

    // Handle bundled files
    if data == nil {
      // Handle both "file.json" and "file"
      if fileString.hasSuffix(".json") {
        let url = URL(string: file)
        fileString = url!.deletingPathExtension().absoluteString
      }
      
      let filePath = Bundle.main.url(forResource: fileString, withExtension: "json")
      data = try? Data(contentsOf: filePath!)
      
      if data == nil {
        NSLog("[ERROR] The specified file \"\(file)\" could not be loaded. Trying absolute path")
        data = try? Data(contentsOf: URL(string: file)!)
        
        if data == nil {
          NSLog("[ERROR] The specified file \"\(file)\" could not be loaded again. Trying absolute path")
          return nil
        } else {
          NSLog("[INFO] Loading file \(file)")
        }
      }
    }
    
    return try? JSONSerialization.jsonObject(with: data!) as? [String: Any]
  }
  
  func initializeView() {
    let _ = animationView()
  }

  public override func frameSizeChanged(_ frame: CGRect, bounds: CGRect) {
    super.frameSizeChanged(frame, bounds: bounds)
    if let animationView = _animationView {
      animationView.frame = bounds
    }
  }
}

// MARK: Public APIs

extension TiAnimationAnimationView {
  
  func play(with completionHandler: KrollCallback?) {
    animationView().play { animationFinished in
      self.processCompleteEvent(with: completionHandler, animationFinished: animationFinished)
    }
  }
  
  func play(_ fromFrame: Float, toFrame: Float, completionHandler: KrollCallback?) {
    animationView().play(fromFrame: AnimationFrameTime(fromFrame),
                         toFrame: AnimationFrameTime(toFrame)) { animationFinished in
      self.processCompleteEvent(with: completionHandler, animationFinished: animationFinished)
    }
  }
  
  // TODO: Expose to module in next major, because we need to break the 3 parameters into a dictionary to be more flexible
  func play(_ fromProgress: Float, toProgress: Float, completionHandler: KrollCallback?) {
    animationView().play(fromProgress: AnimationProgressTime(fromProgress),
                         toProgress: AnimationProgressTime(fromProgress)) { animationFinished in
      self.processCompleteEvent(with: completionHandler, animationFinished: animationFinished)
    }
  }
  
  func pause() {
    animationView().pause()
  }
  
  func stop() {
    animationView().stop()
  }
  
  func add(view: UIView, toLayer: String) {
    let animationSubview = AnimationSubview(frame: view.frame)
    animationSubview.addSubview(view)
    
    animationView().addSubview(animationSubview, forLayerAt: AnimationKeypath(keypath: toLayer))
  }
  
  func isPlaying() -> Bool {
    animationView().isAnimationPlaying
  }
  
  func duration() -> Float {
    NSLog("[ERROR] The \"duration\" property has been removed from the native library")
    return 0.0
  }
  
  func frame() -> Float {
    return Float(animationView().currentFrame)
  }
  
  func progress() -> Float {
    return Float(animationView().currentProgress)
  }
  
  func speed() -> Float {
    return Float(animationView().animationSpeed)
  }
  
  func loop() -> Bool {
    return animationView().loopMode == .loop
  }
  
  func setProgress(progress: Float) {
    animationView().currentProgress = CGFloat(progress)
  }
  
  func setSpeed(speed: Float) {
    animationView().animationSpeed = CGFloat(speed)
  }
  
  func setLoop(loop: Bool) {
    animationView().loopMode = loop ? .loop : .playOnce
  }
  
  func setCacheEnabled(cacheEnabled: Bool) {
    NSLog("[ERROR] The \"cache\" property has been removed from the native library")
  }
  
  func cacheEnabled() -> Bool {
    NSLog("[ERROR] The \"cache\" property has been removed from the native library")
    
    return false
  }
}
